### PR TITLE
chore: conversation filter use case - WPB-7301

### PR DESCRIPTION
**Pull request by caldrian**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-7301" title="WPB-7301" target="_blank">StoryWPB-7301</a>  [iOS] Search: "Conversations" - Add search bar to top of conversation list
  </td></table>
  <br />
 

### Issue

This PR adds a use case and protocols as a preparation to implement conversation searching/filtering in the UI.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [x] Make sure you use the API for UI elements that support large fonts.
- [x] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [x] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

File: wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireUtilities
+
+/// This use case takes grouped conversations as input and returns only non-empty groups of conversations
+/// where the provided search text matches.
+///
+/// The search text is compared to the name of conversations as well as the names of the conversation's participants.
+public struct SearchConversationsUseCase<ConversationContainer>: SearchConversationsUseCaseProtocol
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift]: The name `SearchConversationsUseCase` might lead some (including myself) to think it's about searching within conversations (i.e their content). Perhaps a less ambiguous name could be  `FilterConversationsUseCase`.

File: wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+where ConversationContainer: MutableConversationContainer {
+
+    private let conversationContainers: [ConversationContainer]
+
+    public init(conversationContainers: [ConversationContainer]) {
+        self.conversationContainers = conversationContainers
+    }
+
+    public func invoke(searchText: String) -> [ConversationContainer] {
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift]:
```suggestion
    public func invoke(query: String) -> [ConversationContainer] {
```

File: wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+/// A mutable container of conversations used by the `ConversationSearchUseCase`.
+/// This can be interpreted as a section in a table where a row is a conversation.
+public protocol MutableConversationContainer {
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift:21]: What do you expect the conforming type to be? Just an array?

File: wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift (hunk 1 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/SearchConversationsUseCase.swift]: Thank you, I've changed it.

File: wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift (hunk 3 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift:21]: It can be anything as long as it satisfies the requirements (`conversations` getter and `removeConversation(at:)` function). 

In the navigation overhaul branch the type `ConversationListViewModel.Section` conforms to this protocol.

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase Protocol.swift [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+/// A use case which returns a collection of grouped conversations which match the provided search text.
+public protocol FilterConversationsUseCaseProtocol {
+    associatedtype ConversationContainer: MutableConversationContainer
+
+    /// Returns conversations which match the `searchText`.
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase Protocol.swift]:
```suggestion
    /// Returns conversations which match the `query`.
```

File: wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift (hunk 3 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift:21]: Could you please clarify why we need this abstraction? I think I'm missing some context.

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                where (participant.searchableName.normalizedForSearch() as String).lowercased().contains(query) {
+                    continue conversationLoop
+                }
+
+                // no match, remove conversation from results
+                conversationContainers[containerIndex].removeConversation(at: conversationIndex)
+            }
+        }
+
+        return conversationContainers
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift]: At first I found this a bit hard to read, but I as I look at it more it makes sense. I'm just wondering if there's a way we can express it in simpler terms, something like:

```swift
return conversationContainers.filterConversations { conversation in
  if conversation.searchableName.contains(query) {
    return true
  }

  if conversation.searchableParticipants.contains { participant in
    participant.searchableName.contains(query)
  } {
    return true
  }
}
```

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        // iterate through all containers and all conversations and remove those who don't match the query
+        var conversationContainers = conversationContainers
+        for containerIndex in conversationContainers.indices {
+        conversationLoop:
+            for conversationIndex in conversationContainers[containerIndex].conversations.indices.reversed() {
+
+                let conversation = conversationContainers[containerIndex].conversations[conversationIndex]
+
+                // check if conversation name matches
+                if (conversation.searchableName.normalizedForSearch() as String).lowercased().contains(query) {
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift]: `searchableName` suggests to me that it's already normalized and lowercased and thus not necessary to do it at the call site.

File: wire-ios-data-model/Source/ConversationSearch/SearchableConversation.swift [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+/// A protocol which represents a conversation that can be searched for.
+public protocol SearchableConversation {
+    associatedtype SearchableParticipant: SearchableConversationParticipant
+
+    var searchableName: String { get }
+    var searchableParticipants: [SearchableParticipant] { get }
+}
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/ConversationSearch/SearchableConversation.swift]: Is the purpose to conceal managed objects from the view layer?

Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/SearchableConversation.swift]: Yes, to make the search algorithm independent from the CoreData entities.

File: wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift (hunk 3 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift:21]: Take a look into `ConversationListViewModel.Section`. It is used as a conversation-grouping model for the conversation list, but has the `fileprivate` access modifier and is not implemented in a clean way, in my opinion.

So I saw three options:
1. Implementing the filter algorithm (`FilterConversationsUseCase`) in Wire-iOS on top of the UI code using the actual types declared there (`ConversationListViewModel` etc.): I actually don't consider it an option.
2. Abstracting the search algorithm using simple collection types (like Array): Then unfortunately a double mapping is required, first from the existing data structure (`ConversationListViewModel.Section`) to an array and after the filtering, mapping the filtered array back to the custom data structure.
3. Declaring the generic `MutableConversationContainer` which allows for getting the current state and removing (= filtering out) items. This is the cleanest approach in my opinion.

Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift:21]: This is the original PR, btw:
https://github.com/wireapp/wire-ios/pull/1605

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift (hunk 6 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift]: Good point, thank you.
Still, I would like the normalization to be a part of the use case. Do you have a suggestion for a better property name?

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift (hunk 5 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift]: I have tried to rewrite the code but failed to find a more compact or easier to read solution, so I just improved the comments.

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift (hunk 6 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift]: I now removed the prefix "searchable".

File: wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved
Comment (samwyndham) [inline wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved:1]: Were changes in this file intentionally committed?

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+        if query.isEmpty {
+            return conversationContainers
+        }
+
+        // Iterate through the grouped conversations and remove the conversations which don't match the query.
+        // Empty containers (conversation groups) will be kept in the result.
+        var containers = conversationContainers
+        for containerIndex in containers.indices {
+        conversationLoop:
```
Comment (samwyndham) [inline wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift:49]: Wow I don't think I've seen this API used in practice. Took me a while to work out what was happening!

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireUtilities
+
+/// This use case takes grouped conversations as input and returns the same groups with their conversations
+/// filtered by the search text `query`.
+///
+/// The search text is compared to the name of conversations as well as the names of the conversation's participants.
+public struct FilterConversationsUseCase<ConversationContainer>: FilterConversationsUseCaseProtocol
```
Comment (samwyndham) [inline wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift:25]: Just sharing a thought - not a request for change... 

Part of me wonders if this should be done at the core data layer (if that is even possible). I guess by doing it here we can have more flexibility.

File: wire-ios-data-model/Tests/ConversationSearch/FilterConversationsUseCaseTests.swift [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        XCTAssertEqual(filtered, conversationContainers)
+    }
+
+    func testSingleCharacterSearchTextReturnsNoConversations() {
+
+        // Given
+        sut = .init(conversationContainers: conversationContainers)
+
+        // When
+        let filtered = sut.invoke(query: "y")
```
Comment (samwyndham) [inline wire-ios-data-model/Tests/ConversationSearch/FilterConversationsUseCaseTests.swift:50]: I don't think the data contains a "y". If you change this to an "a" I think you test will fail right? Maybe I misunderstand the purpose of this test. I thought it means that any single character search will always return no conversations

**Review by samwyndham — APPROVED** [review]

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift (hunk 9 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift:25]: Good question, I don't know. I have little experience with search in general. The way of searching is also not flexible, it just calls String.contains.
The other point why I did this isolated was because the code for showing the conversation list is really hard to maintain. I hope we can refactor the whole feature (conversation list including search) eventually.

File: wire-ios-data-model/Tests/ConversationSearch/FilterConversationsUseCaseTests.swift (hunk 10 above)
Comment (caldrian) [reply wire-ios-data-model/Tests/ConversationSearch/FilterConversationsUseCaseTests.swift:50]: The name was misleading. The test is about not matching anything.
I changed the name.

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift (hunk 8 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift:49]: You mean the label? I need it for continuing the outer loop from within a nested loop.

File: wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved
Comment (caldrian) [reply wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved:1]: No, shall I create a separate PR for these?

File: wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift (hunk 8 above)
Comment (samwyndham) [reply wire-ios-data-model/Source/ConversationSearch/FilterConversationsUseCase.swift:49]: Yes exactly. Nothing wrong with it. I just completely forgot about this feature from swift.

File: WireUI/Sources/WireAccountImageUI/AccountImageGenerator/AccountImageGenerator.swift [hunk 11]
```diff
@@ -61,6 +61,7 @@ private final class InitialsRenderer: UIView {
 
     private func setupLabel(_ initials: String) {
         initialsLabel.font = .systemFont(ofSize: 13, weight: .light)
+        initialsLabel.textColor = .black
```
Comment (agisilaos) [inline WireUI/Sources/WireAccountImageUI/AccountImageGenerator/AccountImageGenerator.swift]: Setting it to black means when switching to dark mode the label's color would still be black. Is this actually the desired behaviour?

**Review by agisilaos — APPROVED** [review]

File: wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift (hunk 3 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift:21]: It seems that without refactoring also the view layer here we are limited in nice solutions and I agree that the one you have chosen is the cleanest option at the moment.

Ideally however I would imagine that the view model requests a simple grouped list of conversations via a use case, passing in an optional filter. I think the filter implementation shouldn't be constrained by the implementation of the view, but rather the view needs to work with a well designed interface provided to it.

For instance:

```swift
let conversationsByCategory: [ConversationCategory: [Conversation]] = usecase.invoke(filter: .query("foo"))
```

**Review by johnxnguyen — APPROVED** [review]

File: wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift (hunk 3 above)
Comment (caldrian) [reply wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift:21]: Yes, this would be nice.
I vote for rewriting the conversation list in SwiftUI eventually.

File: WireUI/Sources/WireAccountImageUI/AccountImageGenerator/AccountImageGenerator.swift (hunk 11 above)
Comment (caldrian) [reply WireUI/Sources/WireAccountImageUI/AccountImageGenerator/AccountImageGenerator.swift]: Good point, I will revert this and ask in the design team.
